Argmax and Argmin Support

umadevimcw opened the pull request:
@muthutt 


- Added argmax support (supports only H and W dimensions only)
- Currently, the code is not yet optimized. Implementation might have repeated modules/modules that can reused with little modification. Will update it once the code is finalized and optimized

umadevimcw committed:
#4215: Add initial version of argmax implementation that support H and W dimension
     : Add argmin support across H, W and HW dimension
     : Refactor test and add assert

muthutt reviewed (approved):
LGTM

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+
+    auto& input_shape = input_a.shape();
+    int size = input_shape[0] * input_shape[1] * input_shape[2] * input_shape[3];
+    TT_ASSERT((input_shape[0] == 1 && input_shape[1] == 1), "Unsupported shapes, supported shapes [1, 1, N, M]");
```
Use TT_FATAL. You should also assert for input tensor rank 4.

umadevimcw replied:
Addressed

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
 
+
+// Argmax returns the index of maximum element in the tensor
+Tensor _argmax(const Tensor& input_a, int dim, const MemoryConfig& output_mem_config) {
```
Similar comments to argmin

umadevimcw replied:
addressed

tt-aho commented on tt_eager/tt_numpy/functions.hpp:
```diff
@@ -189,6 +189,85 @@ static Tensor index_trilu(const Shape& shape, const int32_t diag, DataType data_
     return output;
 }
 
+
+template<typename T>
```
Assert for rank 4 in these functions since your loops are hardcoded.

TODO is to remove the hardcoded dim loops to generically support any dim and possibly refactor into single function (not a requirement for this pr).

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+
+    if (dim == 3 || dim == -1)
+    {
+        Tensor tindex = tt::numpy::index_width<bfloat16>(input_shape, DataType::BFLOAT16);
+        Tensor min_val = reduce(input_a, ReduceOpMath::MIN, ReduceOpDim::W);
+        Tensor min_tensor = zeros_like(input_a, output_mem_config);
+        min_tensor = bcast(min_tensor, min_val,  BcastOpMath::ADD, BcastOpDim::W, output_mem_config);
+        Tensor cmp_results = eq(input_a, min_tensor, std::nullopt, output_mem_config);
+        Tensor min_indices =  mul(cmp_results ,tindex, std::nullopt, output_mem_config);
+        Tensor midx = full_like(min_indices, size);
+        Tensor result = where(eqz(min_indices), midx, min_indices);
+        result = reduce(result, ReduceOpMath::MIN, ReduceOpDim::W);
+        Tensor res_index = zeros_like(result, output_mem_config);
+        result = where(eq(result, full_like(result, size)), res_index, result);
+        res_index = bcast(res_index, result,  BcastOpMath::ADD, BcastOpDim::W, output_mem_config);
+        std::vector<int64_t> permute_dims = {0, 1, 3, 2};
+        Tensor transpose_res = permute(res_index,permute_dims);
```
You should scope/dellocate your tensors to reduce memory usage.
```
         Tensor min_val = reduce(input_a, ReduceOpMath::MIN, ReduceOpDim::W);
         Tensor min_tensor = zeros_like(input_a, output_mem_config);
         min_tensor = bcast(min_tensor, min_val,  BcastOpMath::ADD, BcastOpDim::W, output_mem_config);
         min_val.deallocate();
         Tensor cmp_results = eq(input_a, min_tensor, std::nullopt, output_mem_config);
         min_tensor.deallocate();
         Tensor tindex = tt::numpy::index_width<bfloat16>(input_shape, DataType::BFLOAT16);
         Tensor min_indices =  mul(cmp_results ,tindex, std::nullopt, output_mem_config);
         // tindex not on device
         // tindex.deallocate();
         Tensor midx = full_like(min_indices, size);
         Tensor result = where(eqz(min_indices), midx, min_indices);
         min_indices.deallocate();
         // midx not on device
         // midx.deallocate();
         result = reduce(result, ReduceOpMath::MIN, ReduceOpDim::W);
         Tensor res_index = zeros_like(result, output_mem_config);
         result = where(eq(result, full_like(result, size)), res_index, result);
         res_index = bcast(res_index, result,  BcastOpMath::ADD, BcastOpDim::W, output_mem_config);
         result.deallocate();
         std::vector<int64_t> permute_dims = {0, 1, 3, 2};
         Tensor transpose_res = permute(res_index,permute_dims);
```

umadevimcw replied:
Updated the code

muthutt committed:
#0: argmax, argmin test update
#0: move index_width and index_height and index_triul to work with generalized rank tensor
#0: fix the bug in index_hw, index_h, and index_w
#0: skip for WHB0 until Reems PR merged
#0: fix per review

umadevimcw commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+Tensor _argmin(const Tensor& input_a, int dim, const MemoryConfig& output_mem_config) {
+
+    auto& input_shape = input_a.shape();
+    int size = input_shape[0] * input_shape[1] * input_shape[2] * input_shape[3];
```
@muthutt Updated the code

umadevimcw commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+
+
+// Argmax returns the index of maximum element in the tensor
+Tensor _argmin(const Tensor& input_a, int dim, const MemoryConfig& output_mem_config) {
```
Addressed

umadevimcw commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+        res_index = bcast(res_index, result,  BcastOpMath::ADD, BcastOpDim::H, output_mem_config);
+        return res_index;
+    }
+    else{
```
Currently,  we are supporting H & W dimensions. and we have added a TT_FATAL to check dim 0, 1 should be equal to 1

umadevimcw committed:
#4125: Update argmin and argmax support with assert condition and deallocate
     : assert -> fatal

umadevimcw commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,136 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+        result = where(eq(result, full_like(result, size)), res_index, result);
+        res_index = bcast(res_index, result,  BcastOpMath::ADD, BcastOpDim::W, output_mem_config);
+        std::vector<int64_t> permute_dims = {0, 1, 3, 2};
+        Tensor transpose_res = permute(res_index,permute_dims);
```
Updated by @muthutt

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,170 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+Tensor _argmax(const Tensor& input_a, int64_t _dim, const MemoryConfig& output_mem_config) {
+
+    auto& input_shape = input_a.shape();
+    TT_ASSERT(input_shape.rank() == 4);
```
Use TT_FATAL here

tt-aho commented on tt_eager/tt_dnn/op_library/composite/composite_ops.cpp:
```diff
@@ -1308,6 +1309,170 @@ Tensor repeat(const Tensor& input_a, const Shape& shape_b, const MemoryConfig& o
+Tensor _argmin(const Tensor& input_a, int64_t _dim, const MemoryConfig& output_mem_config) {
+
+    auto& input_shape = input_a.shape();
+    TT_ASSERT(input_shape.rank() == 4);
```
Use TT_FATAL here
